### PR TITLE
Security: Store datasource passwords encrypted in secureJsonData

### LINKED ISSUE #10827: Encrypt Data Source Basic Auth in DB
As part of compliance at the company I work I need to know how Grafana handles passwords and keys. I had a vague look at the code and I can see that usernames and passwords are stored in the database encrypted. However I have noticed that the credentials for each data source is stored unencrypted (as far as I can tell). 

This issue is to confirm whether this is actually the case and to also see if it would be possible to store these credentials encrypted. 

Please include this information:
- What Grafana version are you using?

4.6.3

- What datasource are you using?

Any data source with Basic auth enabled

- What OS are you running grafana on?

MacOS for dev, Linux in Prod

- What did you do?

Added basic auth to data source

- What was the expected result?

data source credentials are stored in database encrypted

- What happened instead?

data source credentials are stored in database unencrypted

### PR DESCRIPTION
**Pull request by aocenas**
Closes https://github.com/grafana/grafana/issues/10827

This got a bit wide in the sense that it needs to touch much more parts than expected. Because of that I would welcome some early feedback if the approach still makes sense.

For data sources we store passwords either in non encrypted columns or in secureJsonData which is encrypted. MSSql and Postgres already store them encrypted other datasources store them plain text. What this PR is doing. Some caveats:
- We store passwords also for browser only datasources which need to be sent back to front end in plaintext.
- We have a provisioning format that already uses either secureJsonData or the plain fields depending on the datasource.
- Non core datasources can already use secureJsonData for storage.

What this PR does so far:
- Adds helper methods on Datasource to return plaintext passwords from secureJsonData if exists with fallback to plaintext db columns.
- Uses this to return plaintext passwords to frontend for browser access datasources.
- ~~Adds migration that will move passwords from plaintext fields to secureJsonData for core datasources.~~
- ~~When core datasource is provisioned, both password or secureJsonData.password can be used and both will save the password in secureJsonData so the format stays the same but the behaviour changes.~~

After talking to other people it will make more sense to not complicate the code with migration and provisioning changes and instead users with existing datasources can migrate by resaving their datasources or altering their provisioned files. New datasources will be created with encrypted passwords by default.

TODO:
- [x] Fix existing tests
- [x] Add unitests
- [x] Manual testing

### CODE REVIEW

File: pkg/api/frontendsettings.go [hunk 1]
```diff
@@ -80,22 +81,21 @@ func (hs *HTTPServer) getFrontendSettingsMap(c *m.ReqContext) (map[string]interf
 
 		if ds.Access == m.DS_ACCESS_DIRECT {
 			if ds.BasicAuth {
-				dsMap["basicAuth"] = util.GetBasicAuthHeader(ds.BasicAuthUser, ds.BasicAuthPassword)
+				dsMap["basicAuth"] = util.GetBasicAuthHeader(ds.BasicAuthUser, ds.DecryptedBasicAuthPassword())
```
Comment (aocenas) [inline pkg/api/frontendsettings.go]: Returns plain text passwords here, no matter if they are saved in secureJsonData or in password/basic_auth_password fields.

File: pkg/cmd/grafana-cli/commands/migrate_database_command.go [hunk 2]
```diff
@@ -0,0 +1,7 @@
+package commands
+
+// migrateDatabaseCommand is just a stub that does nothing. Migrations are run as part of every database command
+// and so we just need that and this can be empty.
+func migrateDatabaseCommand(c CommandLine) error {
```
Comment (aocenas) [inline pkg/cmd/grafana-cli/commands/migrate_database_command.go]: Added a command to just run migration so it is easier to test the migration. Mainly just makes the log output better than running server and finding the logs there.

File: pkg/plugins/plugins.go [hunk 3]
```diff
@@ -237,3 +237,27 @@ func GetPluginMarkdown(pluginId string, name string) ([]byte, error) {
 	}
 	return data, nil
 }
+
+var coreDatasourceTypes = []string{
```
Comment (aocenas) [inline pkg/plugins/plugins.go]: Need to distinguish between core and other datasource. Not sure if there is better way to do this.

File: pkg/services/provisioning/datasources/config_reader.go [hunk 4]
```diff
@@ -51,35 +51,50 @@ func (cr *configReader) parseDatasourceConfig(path string, file os.FileInfo) (*D
 		return nil, err
 	}
 
-	var apiVersion *ConfigVersion
-	err = yaml.Unmarshal(yamlFile, &apiVersion)
+	apiVersion, err := getDatasourceConfigApiVersion(yamlFile)
```
Comment (aocenas) [inline pkg/services/provisioning/datasources/config_reader.go]: This file was just refactored.

File: pkg/services/provisioning/datasources/datasources.go [hunk 5]
```diff
@@ -39,14 +39,14 @@ func (dc *DatasourceProvisioner) apply(cfg *DatasourcesAsConfig) error {
...
 		err := bus.Dispatch(cmd)
-		if err != nil && err != models.ErrDataSourceNotFound {
-			return err
-		}
-
-		if err == models.ErrDataSourceNotFound {
-			dc.log.Info("inserting datasource from configuration ", "name", ds.Name)
-			insertCmd := createInsertCommand(ds)
-			if err := bus.Dispatch(insertCmd); err != nil {
+		if err != nil {
```
Comment (aocenas) [inline pkg/services/provisioning/datasources/datasources.go]: Again just a refactor of the conditions.

File: pkg/services/provisioning/datasources/types.go [hunk 6]
```diff
@@ -116,6 +117,19 @@ func (cfg *DatasourcesAsConfigV1) mapToDatasourceFromConfig(apiVersion int64) *D
 	}
 
 	for _, ds := range cfg.Datasources {
+
+		// To provide better security, core datasources store passwords in secureJsonData but that was not always the
+		// case that is why provisioning files can use password field directly. This normalizes it so both password
+		// field and secureJsonData can be used in config. For non core datasources they have to opt in explicitly to
+		// use secureJsonData.
+		if plugins.IsCoreDatasource(ds.Type) {
```
Comment (aocenas) [inline pkg/services/provisioning/datasources/types.go]: Special behaviour for core datasources. Not nice but keeps the format backward compatible.

File: pkg/services/sqlstore/migrations/encrypt_passwords.go [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+		return nil, err
+	}
+
+	jsonFieldName := util.ToCamelCase(passwordFieldName)
+	secureJsonData[jsonFieldName] = encryptedPassword
+	return secureJsonData, nil
+}
+
+//
+// All functions below are copy pasted from encryption.go and encoding.go to keep this migration as stable as possible.
```
Comment (aocenas) [inline pkg/services/sqlstore/migrations/encrypt_passwords.go]: Maybe unnecessary caution here. It is bad practice to link to business logic in migration, but this could be seen as simple util functions that should not change (for example I did not copy `util.ToCamelCase` here)

File: pkg/services/sqlstore/migrations/encrypt_passwords.go [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+	MigrationBase
+}
+
+func (m *AddEncryptPasswordsMigration) Sql(dialect Dialect) string {
+	return "code migration"
+}
+
+func (m *AddEncryptPasswordsMigration) Exec(sess *xorm.Session, mg *Migrator) error {
+	var passwordRows []map[string]string
+	datasourceTypes := []string{
```
Comment (aocenas) [inline pkg/services/sqlstore/migrations/encrypt_passwords.go]: These seems to be the only ones that use plaintext passwords, if there are other please let me know.

File: pkg/services/provisioning/datasources/types.go (hunk 6 above)
Comment (marefr) [reply pkg/services/provisioning/datasources/types.go]: Is this special behaviour really needed given the fallbacks used in `DecryptedBasicAuthPassword` and `DecryptedPassword`?

File: pkg/services/sqlstore/migrations/encrypt_passwords.go (hunk 7 above)
Comment (marefr) [reply pkg/services/sqlstore/migrations/encrypt_passwords.go]: Yes think this is unnecessary caution. Changing these utils will result in no users being able to login anymore so think it's better to handle altogether when that day comes :smile:

File: pkg/services/provisioning/datasources/config_reader.go [hunk 9]
```diff
@@ -51,35 +51,50 @@ func (cr *configReader) parseDatasourceConfig(path string, file os.FileInfo) (*D
...
+	var apiVersion *ConfigVersion
+	err := yaml.Unmarshal(yamlFile, &apiVersion)
 	if err != nil {
-		return nil, err
+		return -1, err
 	}
 
-	cr.log.Warn("[Deprecated] the datasource provisioning config is outdated. please upgrade", "filename", filename)
+	if apiVersion == nil {
+		return 0, nil
```
Comment (marefr) [inline pkg/services/provisioning/datasources/config_reader.go]: Shouldn't this be 1 per default? On another note maybe better to do config_reader refactoring in a separate PR to make this PR smaller?

File: pkg/services/provisioning/datasources/datasources.go (hunk 5 above)
Comment (marefr) [reply pkg/services/provisioning/datasources/datasources.go]: Maybe better to do this refactoring in a separate PR to make this PR smaller?

File: pkg/services/sqlstore/migrations/encrypt_passwords.go [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+		return fmt.Errorf("basic_auth_password updates failed: %v", err)
+	}
+
+	return nil
+}
+
+// addEncryptPasswordsMigration will move unencrypted passwords from password and basic_auth_password fields into
+// secure_json_data and encrypt them. This is done only for some core datasources that did not use the encrypted storage
+// until now.
+func addEncryptPasswordsMigration(mg *Migrator) {
```
Comment (marefr) [inline pkg/services/sqlstore/migrations/encrypt_passwords.go]: Better naming of this, the file and struct, like `DatasourcePasswordMigration` to denote it's a migration for datasources? The code can even be part of datasource_mig.go to put all migrations related to datasource in same file? Then you don't need to add the migration from migrations.go and can do it from datasource_mig.go instead.

File: pkg/services/provisioning/datasources/types.go (hunk 6 above)
Comment (marefr) [reply pkg/services/provisioning/datasources/types.go]: And also shouldn't this logic be applied when creating/updating new/existing datasources through api as well?

**Review by marefr — COMMENTED** [review]
Have left a few comments. Think this is a good start. Haven't tested it yet though.

File: pkg/services/provisioning/datasources/datasources.go (hunk 5 above)
Comment (aocenas) [reply pkg/services/provisioning/datasources/datasources.go]: Yes you are right will do that separately.

File: pkg/services/provisioning/datasources/types.go (hunk 6 above)
Comment (aocenas) [reply pkg/services/provisioning/datasources/types.go]: > Is this special behaviour really needed given the fallbacks used in DecryptedBasicAuthPassword and DecryptedPassword?

Not needed but it would mean that you will have passwords plaintext until you manually change your provisioning config. Also as we already have provisioning which use both `password` field and `secureJsonData: { password }` depending on the data source and I am not sure if it makes sense to break these. Like we could create a new version where you would just use `password` that would be automatically encrypted and store in `secureJsonData`.

> And also shouldn't this logic be applied when creating/updating new/existing datasources through api as well?

Yes I did not thought about that to be honest but probably should be in sync to some extent.

File: pkg/services/sqlstore/migrations/encrypt_passwords.go (hunk 10 above)
Comment (aocenas) [reply pkg/services/sqlstore/migrations/encrypt_passwords.go]: > The code can even be part of datasource_mig.go to put all migrations related to datasource in same file? Then you don't need to add the migration from migrations.go and can do it from datasource_mig.go instead.

Not sure about this. It seemed weird to me that migrations are added out of order and talking to @bergquist he suggested that going forward they should be ordered. Or do you mean just having new migration function in `datasource_mig.go` but calling it in correct order afterwards in `migration.go`?

File: pkg/services/provisioning/datasources/types.go (hunk 6 above)
Comment (marefr) [reply pkg/services/provisioning/datasources/types.go]: > Not needed but it would mean that you will have passwords plaintext until you manually change your provisioning config.

Wouldn't the database migration take care of that? Or is the migration only updating certain datasource types and not all?

File: pkg/services/sqlstore/migrations/encrypt_passwords.go (hunk 10 above)
Comment (marefr) [reply pkg/services/sqlstore/migrations/encrypt_passwords.go]: Okay wasn't aware of that suggestion. For me it should be enough that they're ordered in each file or am I missing something @bergquist? Makes it easier to find migrations related to same table.

Comment (aocenas) [reply pkg/services/sqlstore/migrations/encrypt_passwords.go]: > For me it should be enough that they're ordered in each file

Ach did not realise that is the case actually. Wouldn't there be some cases where that could still break? Like merging some fields from 2 different tables or some data migration based on some values from different table? I always worked with migration system where each migrations was somehow uniquely marked and ordered and append only so that is why this seems strange to me and wonder if there could be some issues from that.

File: pkg/services/provisioning/datasources/types.go (hunk 6 above)
Comment (aocenas) [reply pkg/services/provisioning/datasources/types.go]: Migration will only change what ever is in DB right now so I wanted to make sure this holds:
`old datasource + migration = new datasource` if that makes sense, and that mean we need to migrate all datasources in the DB and have a new provisioning.
But maybe this is not needed?
This all is done only for core datasources so the premise of this PR is that you can still use both encrypted or plaintext storage for password. Maybe it would be enough to just say `"change your provisioning files and reset your passwords after the update if you want passwords to be encrypted"` in the change log and so the migration would not be needed?

File: pkg/services/provisioning/datasources/config_reader.go (hunk 9 above)
Comment (aocenas) [reply pkg/services/provisioning/datasources/config_reader.go]: As I mentioned will do a separate PR with this but I think this is correct, in the previous code it was:
```go
    if apiVersion == nil {
		apiVersion = &ConfigVersion{ApiVersion: 0}
	}

	if apiVersion.ApiVersion > 0 {
		var v1 *DatasourcesAsConfigV1
```
so nil => 0, 0 => 0, > 0 => 1

File: pkg/cmd/grafana-cli/commands/migrate_database_command.go (hunk 2 above)
Comment (bergquist) [reply pkg/cmd/grafana-cli/commands/migrate_database_command.go]: Great! Wish this would have been a separate PR :) but it doesn't matter now.

File: pkg/services/provisioning/datasources/types.go (hunk 6 above)
Comment (marefr) [reply pkg/services/provisioning/datasources/types.go]: Ah okay my bad! Thought we we're migrating "all" datasources. In that case I think this is okay during migration phase. Think we should question when we're feeling sort of done with this PR whether it would make sense to migrate all datasources or not given that the code may be easier to reason about and/or decide when we migrate all the other datasources.

File: pkg/services/provisioning/datasources/config_reader.go (hunk 9 above)
Comment (marefr) [reply pkg/services/provisioning/datasources/config_reader.go]: Oh okay :+1:

File: pkg/services/sqlstore/migrations/encrypt_passwords.go (hunk 10 above)
Comment (aocenas) [reply pkg/services/sqlstore/migrations/encrypt_passwords.go]: As the migration was removed, not an issue right now. We can probably have a discussion in backend weekly at some point whether to change the migrations or not.

File: docs/sources/installation/upgrading.md [hunk 11]
```diff
@@ -123,7 +123,18 @@ If you're using systemd and have a large amount of annotations consider temporar
 If you have text panels with script tags they will no longer work due to a new setting that per default disallow unsanitized HTML.
 Read more [here](/installation/configuration/#disable-sanitize-html) about this new setting.
 
-### Authentication and security
+## Upgrading to v6.x
```
Comment (aocenas) [inline docs/sources/installation/upgrading.md]: @torkelo mentioned it would be nice to add something here. Not sure though if it makes sense to include it in this PR or add it later when creating a build

File: docs/sources/administration/provisioning.md [hunk 12]
```diff
@@ -184,8 +188,8 @@ Secure json data is a map of settings that will be encrypted with [secret key](/
 | tlsCACert | string | *All* |CA cert for out going requests |
 | tlsClientCert | string | *All* |TLS Client cert for outgoing requests |
 | tlsClientKey | string | *All* |TLS Client key for outgoing requests |
-| password | string | PostgreSQL | password |
-| user | string | PostgreSQL | user |
```
Comment (aocenas) [inline docs/sources/administration/provisioning.md]: Seems like this was an error. User is taken from user field not from secureJsonData

File: pkg/services/provisioning/datasources/types.go [hunk 13]
```diff
@@ -135,6 +137,12 @@ func (cfg *DatasourcesAsConfigV1) mapToDatasourceFromConfig(apiVersion int64) *D
 			Editable:          ds.Editable,
 			Version:           ds.Version,
 		})
+		if ds.Password != "" {
```
Comment (aocenas) [inline pkg/services/provisioning/datasources/types.go]: Added warnings here, thought not sure if there won't be too much of them if somebody has a lots of datasources.

File: pkg/services/sqlstore/datasource.go [hunk 14]
```diff
@@ -178,6 +178,10 @@ func UpdateDataSource(cmd *m.UpdateDataSourceCommand) error {
 		sess.UseBool("basic_auth")
 		sess.UseBool("with_credentials")
 		sess.UseBool("read_only")
+		// Make sure password are zeroed out if empty. We do this as we want to migrate passwords from
+		// plain text fields to SecureJsonData.
+		sess.MustCols("password")
```
Comment (aocenas) [inline pkg/services/sqlstore/datasource.go]: Interestingly at the moment it is not actually possible to remove values from config. I needed to add these to be able to do that for plain text password, but otherwise no other value can be removed. Probably not a big issue but wondered if how to properly handle such things. Should there be something like `fieldSet` bool when serialising the request? Right now we send the whole object to the API so this should not be an issue, but this won't support a partial PUT requests.

File: public/app/plugins/datasource/influxdb/module.ts [hunk 15]
```diff
@@ -1,8 +1,21 @@
...
+} from '../../../features/datasources/utils/passwordHandlers';
 
 class InfluxConfigCtrl {
   static templateUrl = 'partials/config.html';
+  current: any;
+  onPasswordReset: ReturnType<typeof createResetHandler>;
+  onPasswordChange: ReturnType<typeof createChangeHandler>;
+
+  constructor() {
+    this.onPasswordReset = createResetHandler(this, PasswordFieldEnum.Password);
```
Comment (aocenas) [inline public/app/plugins/datasource/influxdb/module.ts]: Wanted to reuse the handler logic as it is the same everywhere but my Angular is rusty so this patter was the best I come up with.

File: docs/sources/installation/upgrading.md (hunk 11 above)
Comment (marefr) [reply docs/sources/installation/upgrading.md]: Either add it here or create a separate issue for that doesn't matter.

Comment (marefr) [reply docs/sources/installation/upgrading.md]: Think here makes sense

Comment (aocenas) [reply docs/sources/installation/upgrading.md]: In that case is there some release TODO where I can add "Change v6.x to current release in upgrading.md" or something like that?

Comment (marefr) [reply docs/sources/installation/upgrading.md]: Not sure I follow. You want to change the header Upgrading to v6.x? This is targetet for 6.2 milestone so use Upgrading to v6.2.

Comment (aocenas) [reply docs/sources/installation/upgrading.md]: Hmm yes that makes sense though if it does not make it into 6.2 it will have to be changed to 6.3 :). So just wondered if there is some place to track it or maybe some label for the PR to use, marking that there is some release version dependant stuff that needs to be updated in case the release changes.

Comment (marefr) [reply docs/sources/installation/upgrading.md]: You have a good point. Normally we happen to spot that upgrade instructions are needed in the end of a release cycle :smile:  

In this case I would create an issue for "Update upgrade notes" and put in 6.2 milestone and this is something we'll do as a task before release similar to writing What's new in and blog post. So to not lose this text I would suggest opening a separate PR targeting new "Update upgrade notes" issue.

Sounds good?

Comment (aocenas) [reply docs/sources/installation/upgrading.md]: Not sure if exactly what you mean but I created issue https://github.com/grafana/grafana/issues/16524 in the 6.2 milestone to remind us about this and changed the version here to 6.2 as that I would say is most likely. I think this should be enough to not forget about this.

Comment (marefr) [reply docs/sources/installation/upgrading.md]: Sound good :+1:

File: pkg/services/provisioning/datasources/config_reader.go [hunk 16]
```diff
@@ -62,7 +62,7 @@ func (cr *configReader) parseDatasourceConfig(path string, file os.FileInfo) (*D
 	}
 
 	if apiVersion.ApiVersion > 0 {
-		var v1 *DatasourcesAsConfigV1
+		v1 := &DatasourcesAsConfigV1{log: cr.log}
 		err = yaml.Unmarshal(yamlFile, &v1)
```
[the remark this answers is not in the document]
Comment (aocenas) [reply pkg/services/provisioning/datasources/config_reader.go]: Hmm interesting, you are right. I think it is possible yaml does some reflection magic traversing the pointer until it gets to something it can fill. Will fix that anyway.

File: pkg/util/strings.go [hunk 17]
```diff
@@ -66,3 +67,19 @@ func GetAgeString(t time.Time) string {
 
 	return "< 1m"
 }
+
+// ToCamelCase changes kebab case, snake case or mixed strings to camel case. See unit test for examples.
+func ToCamelCase(str string) string {
```
[the remark this answers is not in the document]
Comment (aocenas) [reply pkg/util/strings.go]: Right, was used in the migration and forgot to remove afterwards. Btw isn't there any lint/check for unused code in the testing pipeline?

[the remark this answers is not in the document]
Comment (bergquist) [reply pkg/util/strings.go]: Should be. lets look into that afterward. We dont want unused code hanging around.

[the remark this answers is not in the document]
Comment (aocenas) [reply pkg/util/strings.go]: Maybe because it wasn't triggered because it was used in the tests.

**Review by bergquist — APPROVED** [review]
LGTM!
